File: Assets/Bundles/Shared/Resources/TS/Network/RemoteEvent.lua
-- Compiled with unity-ts v2.1.0-75
local _NetworkAPI = require("Shared/TS/Network/NetworkAPI")
local NetworkAPI = _NetworkAPI.default
local NetworkChannel = _NetworkAPI.NetworkChannel
local ID_COUNTER = 0
local RemoteEventServer
do
	RemoteEventServer = setmetatable({}, {
		__tostring = function()
			return "RemoteEventServer"
		end,
	})
	RemoteEventServer.__index = RemoteEventServer
	function RemoteEventServer.new(...)
		local self = setmetatable({}, RemoteEventServer)
		return self:constructor(...) or self
	end
	function RemoteEventServer:constructor(id, channel)
		if channel == nil then
			channel = NetworkChannel.Reliable
		end
		self.id = id
		self.channel = channel
	end
	function RemoteEventServer:FireAllClients(...)
		local args = { ... }
		NetworkAPI.fireAllClients(self.id, args, self.channel)
	end
	function RemoteEventServer:FireExcept(ignoredClientId, ...)
		local args = { ... }
		NetworkAPI.fireExcept(self.id, ignoredClientId, args, self.channel)
	end
	function RemoteEventServer:FireClient(clientId, ...)
		local args = { ... }
		NetworkAPI.fireClient(self.id, clientId, args, self.channel)
	end
	function RemoteEventServer:FireClients(clientIds, ...)
		local args = { ... }
		NetworkAPI.fireClients(self.id, clientIds, args, self.channel)
	end
	function RemoteEventServer:OnClientEvent(callback)
		return NetworkAPI.connect(self.id, callback)
	end
end
local RemoteEventClient
do
	RemoteEventClient = setmetatable({}, {
		__tostring = function()
			return "RemoteEventClient"
		end,
	})
	RemoteEventClient.__index = RemoteEventClient
	function RemoteEventClient.new(...)
		local self = setmetatable({}, RemoteEventClient)
		return self:constructor(...) or self
	end
	function RemoteEventClient:constructor(id, channel)
		if channel == nil then
			channel = NetworkChannel.Reliable
		end
		self.id = id
		self.channel = channel
	end
	function RemoteEventClient:FireServer(...)
		local args = { ... }
		NetworkAPI.fireServer(self.id, args, self.channel)
	end
	function RemoteEventClient:OnServerEvent(callback)
		return NetworkAPI.connect(self.id, callback)
	end
end
local RemoteEvent
do
	RemoteEvent = setmetatable({}, {
		__tostring = function()
			return "RemoteEvent"
		end,
	})
	RemoteEvent.__index = RemoteEvent
	function RemoteEvent.new(...)
		local self = setmetatable({}, RemoteEvent)
		return self:constructor(...) or self
	end
	function RemoteEvent:constructor(channel)
		if channel == nil then
			channel = NetworkChannel.Reliable
		end
		local _original = ID_COUNTER
		ID_COUNTER += 1
		local id = _original
		self.Server = RemoteEventServer.new(id, channel)
		self.Client = RemoteEventClient.new(id, channel)
	end
end
return {
	RemoteEvent = RemoteEvent,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Bundles/@Easy/Core/Core~/src/Shared/Network/RemoteEvent.ts
import { Player } from "Shared/Player/Player";
import NetworkAPI, { NetworkChannel } from "./NetworkAPI";
import { RemoteKeyHasher } from "./RemoteKeyHasher";

type RemoteParamsToClient<T> = Parameters<
	T extends unknown[]
		? (player: Player, ...args: T) => void
		: T extends unknown
			? (player: Player, arg: T) => void
			: (player: Player) => void
>;

type RemoteParamsToServer<T> = Parameters<
	T extends unknown[] ? (...args: T) => void : T extends unknown ? (arg: T) => void : () => void
>;

type RemoteParamsToAllClients<T> = RemoteParamsToServer<T>;

type RemoteCallbackFromClient<T> = (...args: RemoteParamsToClient<T>) => void;
type RemoteCallbackFromServer<T> = (...args: RemoteParamsToServer<T>) => void;

let ID_COUNTER = 0;

const packageMap = new Map<number, number>();

class RemoteEventServer<T extends unknown[] | unknown> {
	constructor(
		private readonly id: number,
		private readonly channel: NetworkChannel = NetworkChannel.Reliable,
	) {}

	public FireAllClients(...args: RemoteParamsToAllClients<T>) {
		NetworkAPI.fireAllClients(this.id, args, this.channel);
	}

	public FireExcept(ignorePlayer: Player, ...args: RemoteParamsToAllClients<T>) {
		NetworkAPI.fireExcept(this.id, ignorePlayer, args, this.channel);
	}

	public FireClient(player: Player, ...args: RemoteParamsToAllClients<T>) {
		NetworkAPI.fireClient(this.id, player, args, this.channel);
	}

	public FireClients(players: Player[], ...args: RemoteParamsToAllClients<T>) {
		NetworkAPI.fireClients(this.id, players, args, this.channel);
	}

	public OnClientEvent(callback: RemoteCallbackFromClient<T>) {
		return NetworkAPI.connect(true, this.id, callback);
	}
}

class RemoteEventClient<T extends unknown[] | unknown> {
	constructor(
		private readonly id: number,
		private readonly channel: NetworkChannel = NetworkChannel.Reliable,
	) {}

	public FireServer(...args: RemoteParamsToServer<T>) {
		NetworkAPI.fireServer(this.id, args, this.channel);
	}

	public OnServerEvent(callback: RemoteCallbackFromServer<T>) {
		return NetworkAPI.connect(false, this.id, callback);
	}
}

export class RemoteEvent<T extends unknown[] | unknown> {
	public readonly server: RemoteEventServer<T>;
	public readonly client: RemoteEventClient<T>;

	/**
	 *
	 * @param channel
	 * @param packageOffset Temporary workaround param.
	 */
	constructor(remoteIdentifier: string, channel: NetworkChannel = NetworkChannel.Reliable) {
		let id = 0;
		const context = RemoteKeyHasher.GetCallerContext();
		if (context) {
			id = RemoteKeyHasher.GetRemoteHash(context, remoteIdentifier, "_e");
		} else {
			warn(
				`Could not generate id for remote: ${remoteIdentifier}. Unable to determine the context that it was created in. 
				This may result in unexpected network behavior.`,
			);
		}

		this.server = new RemoteEventServer(id, channel);
		this.client = new RemoteEventClient(id, channel);
	}
}
